File: app/src/main/java/com/example/cinemates/model/Person.java
package com.example.cinemates.model;

import androidx.room.Entity;
import androidx.room.Ignore;
import androidx.room.PrimaryKey;
import androidx.room.TypeConverters;

import com.example.cinemates.util.Converters;

import java.io.Serializable;
import java.util.Objects;

/**
 * @author Antonio Di Nuzzo
 * Created 31/05/2022 at 11:32
 */
@Entity
public class Person implements Serializable {
    private String name, profile_path;
    @PrimaryKey
    private Integer id;
    private boolean favorite;
    @TypeConverters(Converters.class)
    private Number popularity;

    @Ignore
    public Person(String name, String profile_path, Integer id, Number popularity) {
        this.name = name;
        this.profile_path = profile_path;
        this.id = id;
        this.popularity = popularity;
    }

    public Person(String name, String profile_path, Integer id, Number popularity, boolean favorite) {
        this.name = name;
        this.profile_path = profile_path;
        this.id = id;
        this.popularity = popularity;
        this.favorite = favorite;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getProfile_path() {
        return profile_path;
    }

    public void setProfile_path(String profile_path) {
        this.profile_path = profile_path;
    }

    public boolean isFavorite() {
        return favorite;
    }

    public void setFavorite(boolean favorite) {
        this.favorite = favorite;
    }

    public void setFavorite() {
        this.favorite = !this.favorite;
    }

    public Integer getId() {
        return id;
    }

    public void setId(Integer id) {
        this.id = id;
    }

    public Number getPopularity() {
        return popularity;
    }

    public void setPopularity(Number popularity) {
        this.popularity = popularity;
    }


    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Person)) return false;
        Person person = (Person) o;
        return Objects.equals(getName(), person.getName()) && getId().equals(person.getId());
    }

    @Override
    public int hashCode() {
        return Objects.hash(getName(), getId());
    }

    @Override
    public String toString() {
        return "Person{" +
                "name='" + name + '\'' +
                ", profile_path='" + profile_path + '\'' +
                ", id=" + id +
                ", popularity=" + popularity +
                '}';
    }
}

File: app/src/main/java/com/example/cinemates/model/Person.kt
package com.example.cinemates.model

import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.PrimaryKey
import java.io.Serializable
import java.time.LocalDate
import java.time.Period
import java.time.format.DateTimeFormatter
import java.util.stream.Collectors

open class Person(
    val adult: Boolean,
    val also_known_as: List<String>,
    val biography: String?,
    val birthday: String?,
    val deathday: String?,
    val gender: Int?,
    val homepage: String?,
    val id: Int,
    val imdb_id: String?,
    val known_for_department: String,
    val name: String,
    val place_of_birth: String?,
    val popularity: Double,
    val profile_path: String?,
) : Serializable {

    val knownAs: String
        get() = also_known_as.stream().collect(Collectors.joining(" - "))

    val age: String
        get() {
            return if (deathday == null && birthday!=null) {
                val birthdayDate = LocalDate.parse(birthday)
                val age = Period.between(
                    birthdayDate,
                    LocalDate.now()
                ).years
                age.toString()
            } else if (birthday != null) {
                val birthdayDate = LocalDate.parse(birthday)
                val deathDayDate = LocalDate.parse(deathday)
                val age = Period.between(
                    birthdayDate,
                    deathDayDate
                ).years
                age.toString()
            }else ""
        }
    val formattedBirthday: String
        get() {
            return if (birthday != null) {
                val localDate =
                    LocalDate.parse(birthday, DateTimeFormatter.ofPattern("yyyy-MM-dd"))
                val pattern = DateTimeFormatter.ofPattern("dd MMM, yyyy")
                pattern.format(localDate)
            } else ""
        }

    val formattedDeathday: String
        get() {
            return if (deathday != null) {
                val localDate =
                    LocalDate.parse(deathday, DateTimeFormatter.ofPattern("yyyy-MM-dd"))
                val pattern = DateTimeFormatter.ofPattern("dd MMM, yyyy")
                pattern.format(localDate)
            } else ""
        }

    constructor(
        adult: Boolean,
        gender: Int?,
        id: Int,
        known_for_department: String,
        name: String,
        popularity: Double,
        profile_path: String?
    ) : this(
        adult, listOf(), null, "", "", gender, null, id, null,
        known_for_department, name, null, popularity, profile_path
    )

    constructor(
        id: Int,
        name: String,
        profile_path: String?,
    ) : this(
        false, arrayListOf(), null, null, null, null, null, id, null,
        "", name, null, 0.0, profile_path
    )

    constructor(
        adult: Boolean,
        gender: Int?,
        id: Int,
        known_for_department: String,
        name: String,
        popularity: Double,
        profile_path: String?,
        also_known_as: List<String>
    ) : this(adult, gender, id, known_for_department, name, popularity, profile_path)


    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Person) return false
        return id == other.id
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }

    override fun toString(): String {
        return "Person(birthday=$birthday, deathDay=$deathday, gender=$gender, id=$id, name='$name', profile_path=$profile_path)"
    }

}